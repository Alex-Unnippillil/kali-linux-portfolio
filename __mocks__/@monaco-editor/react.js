import React from 'react';
export default function MonacoMock({ value = '', onChange }) {
  return React.createElement('textarea', {
    value,
    onChange: e => onChange && onChange(e.target.value),
  });
}
