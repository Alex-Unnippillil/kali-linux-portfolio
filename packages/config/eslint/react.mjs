import globals from 'globals';
import { baseConfig, compat } from './base.mjs';

const reactConfig = [
  ...baseConfig,
  {
    languageOptions: {
      globals: globals.browser,
    },
    rules: {
      'no-top-level-window/no-top-level-window-or-document': 'off',
      'no-undef': 'off',
      'no-unused-vars': 'off',
      'react/react-in-jsx-scope': 'off',
    },
  },
  ...compat.config({
    extends: [
      'eslint:recommended',
      'plugin:@typescript-eslint/recommended',
      'plugin:react/recommended',
      'plugin:react-hooks/recommended',
    ],
    settings: {
      react: {
        version: 'detect',
      },
    },
  }),
  {
    rules: {
      '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
      '@typescript-eslint/no-unused-expressions': ['error', { allowShortCircuit: true }],
      'react/react-in-jsx-scope': 'off',
    },
  },
];

export default reactConfig;
