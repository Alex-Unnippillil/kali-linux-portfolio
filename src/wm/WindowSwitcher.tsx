import React, { useState, useEffect } from 'react';
import keybindingManager from './keybindingManager';
import Icon from '@/components/common/Icon';

export interface WindowInfo {
  id: string;
  title: string;
  icon: string;
  preview?: string;
}

interface Props {
  windows: WindowInfo[];
  onSelect?: (id: string) => void;
}

export default function WindowSwitcher({ windows, onSelect }: Props) {
  const [visible, setVisible] = useState(false);
  const [index, setIndex] = useState(0);
  const [cycleAll, setCycleAll] = useState(false);

  useEffect(() => {
    const handleAltTab = () => {
      setVisible(true);
      setIndex((i) => (i + 1) % windows.length);
    };

    keybindingManager.register('Alt+Tab', handleAltTab);

    const handleKeyUp = (e: KeyboardEvent) => {
      if (visible && e.key === 'Alt') {
        setVisible(false);
        onSelect?.(windows[index]?.id);
        setIndex(0);
      }
    };

    window.addEventListener('keyup', handleKeyUp);

    return () => {
      keybindingManager.unregister('Alt+Tab', handleAltTab);
      window.removeEventListener('keyup', handleKeyUp);
    };
  }, [visible, windows, index, onSelect]);

  if (!visible) return null;

  return (
    <div
      className="window-switcher-overlay"
      style={{
        position: 'fixed',
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        background: 'rgba(0,0,0,0.6)',
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        justifyContent: 'center',
        zIndex: 1000,
      }}
    >
      <div
        className="window-switcher-list"
        style={{ display: 'flex', gap: '1rem' }}
      >
        {windows.map((win, i) => (
          <div
            key={win.id}
            className={
              i === index ? 'window-switcher-item active' : 'window-switcher-item'
            }
            style={{
              padding: '0.5rem',
              border: i === index ? '2px solid #fff' : '2px solid transparent',
              background: '#222',
              color: '#fff',
              width: '200px',
              textAlign: 'center',
            }}
          >
            <Icon src={win.icon} alt="" width={32} height={32} />
            <div>{win.title}</div>
            {win.preview && (
              <img
                src={win.preview}
                alt="preview"
                width={200}
                height={112}
                style={{ width: '100%', marginTop: '0.5rem' }}
              />
            )}
          </div>
        ))}
      </div>
      <label
        className="cycle-all"
        style={{ marginTop: '1rem', color: '#fff' }}
      >
        <input
          type="checkbox"
          aria-label="Cycle through all workspaces"
          checked={cycleAll}
          onChange={() => setCycleAll((v) => !v)}
        />{' '}
        Cycle through all workspaces
      </label>
    </div>
  );
}
