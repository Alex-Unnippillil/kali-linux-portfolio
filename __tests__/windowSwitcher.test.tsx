import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import WindowSwitcher from '../components/screen/window-switcher';

describe('WindowSwitcher', () => {
  const baseWindows = [
    {
      id: 'terminal',
      title: 'Terminal',
      icon: './themes/Yaru/apps/bash.png',
    },
    {
      id: 'firefox',
      title: 'Firefox',
      icon: '/themes/Yaru/apps/firefox.svg',
    },
  ];

  it('renders window icons alongside titles', () => {
    const handleSelect = jest.fn();
    const { asFragment, getByAltText } = render(
      <WindowSwitcher windows={baseWindows} onSelect={handleSelect} />
    );

    expect(getByAltText('Terminal icon')).toBeInTheDocument();
    expect(getByAltText('Firefox icon')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('retains icon markup when filtering results', () => {
    const { asFragment, getByPlaceholderText } = render(
      <WindowSwitcher windows={baseWindows} />
    );

    fireEvent.change(getByPlaceholderText('Search windows'), {
      target: { value: 'term' },
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
