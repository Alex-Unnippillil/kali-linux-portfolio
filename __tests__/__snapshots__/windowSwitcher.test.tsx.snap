// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`WindowSwitcher renders window icons alongside titles 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 z-50 flex items-center justify-center bg-black/70 backdrop-blur-sm"
  >
    <div
      class="w-11/12 max-w-lg rounded-lg border border-black/20 bg-white/95 p-4 text-kali-text shadow-2xl dark:border-white/10 dark:bg-ub-grey dark:text-white md:w-1/3"
    >
      <label
        class="sr-only"
        for="window-switcher-search"
      >
        Search windows
      </label>
      <input
        aria-label="Search windows"
        class="mb-4 w-full rounded-md border border-black/10 bg-white/40 px-3 py-2 text-sm text-kali-text placeholder:text-kali-muted focus:border-kali-focus focus:outline-none focus:ring-2 focus:ring-kali-focus dark:border-white/10 dark:bg-black/40 dark:text-white dark:placeholder:text-gray-300"
        id="window-switcher-search"
        placeholder="Search windows"
        value=""
      />
      <ul
        aria-label="Open windows"
        class="space-y-2"
        role="listbox"
      >
        <li
          aria-selected="true"
          class="group flex items-center gap-3 rounded-md px-3 py-2 text-sm transition-colors bg-ub-orange text-black shadow-lg dark:text-black"
          id="window-switcher-item-terminal"
          role="option"
        >
          <img
            alt="Terminal icon"
            class="h-9 w-9 flex-shrink-0 rounded-md border border-black/10 bg-white/70 object-contain p-1 dark:border-white/10 dark:bg-black/40"
            src="/themes/Yaru/apps/bash.png"
          />
          <span
            class="flex-1 truncate font-medium"
            title="Terminal"
          >
            Terminal
          </span>
        </li>
        <li
          aria-selected="false"
          class="group flex items-center gap-3 rounded-md px-3 py-2 text-sm transition-colors bg-white/30 text-kali-text hover:bg-white/50 dark:bg-black/40 dark:text-white dark:hover:bg-black/60"
          id="window-switcher-item-firefox"
          role="option"
        >
          <img
            alt="Firefox icon"
            class="h-9 w-9 flex-shrink-0 rounded-md border border-black/10 bg-white/70 object-contain p-1 dark:border-white/10 dark:bg-black/40"
            src="/themes/Yaru/apps/firefox.svg"
          />
          <span
            class="flex-1 truncate font-medium"
            title="Firefox"
          >
            Firefox
          </span>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`WindowSwitcher retains icon markup when filtering results 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 z-50 flex items-center justify-center bg-black/70 backdrop-blur-sm"
  >
    <div
      class="w-11/12 max-w-lg rounded-lg border border-black/20 bg-white/95 p-4 text-kali-text shadow-2xl dark:border-white/10 dark:bg-ub-grey dark:text-white md:w-1/3"
    >
      <label
        class="sr-only"
        for="window-switcher-search"
      >
        Search windows
      </label>
      <input
        aria-label="Search windows"
        class="mb-4 w-full rounded-md border border-black/10 bg-white/40 px-3 py-2 text-sm text-kali-text placeholder:text-kali-muted focus:border-kali-focus focus:outline-none focus:ring-2 focus:ring-kali-focus dark:border-white/10 dark:bg-black/40 dark:text-white dark:placeholder:text-gray-300"
        id="window-switcher-search"
        placeholder="Search windows"
        value="term"
      />
      <ul
        aria-label="Open windows"
        class="space-y-2"
        role="listbox"
      >
        <li
          aria-selected="true"
          class="group flex items-center gap-3 rounded-md px-3 py-2 text-sm transition-colors bg-ub-orange text-black shadow-lg dark:text-black"
          id="window-switcher-item-terminal"
          role="option"
        >
          <img
            alt="Terminal icon"
            class="h-9 w-9 flex-shrink-0 rounded-md border border-black/10 bg-white/70 object-contain p-1 dark:border-white/10 dark:bg-black/40"
            src="/themes/Yaru/apps/bash.png"
          />
          <span
            class="flex-1 truncate font-medium"
            title="Terminal"
          >
            Terminal
          </span>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
